src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseThreadStaticCorrectly.cs

```diff
@@ -0,0 +1,108 @@
+    {
+        internal const string ThreadStaticNonStaticFieldRuleId = "CA2259";
+        internal const string ThreadStaticInitializedInlineRuleId = "CA2019";
+
+        // [ThreadStatic]
+        // private object t_nonStaticField;
+        internal static readonly DiagnosticDescriptor ThreadStaticOnNonStaticFieldRule = DiagnosticDescriptorHelper.Create(ThreadStaticNonStaticFieldRuleId,
+            CreateLocalizableResourceString(nameof(ThreadStaticOnNonStaticFieldTitle)),
+            CreateLocalizableResourceString(nameof(ThreadStaticOnNonStaticFieldMessage)),
+            DiagnosticCategory.Usage,
+            RuleLevel.BuildWarning,
+            CreateLocalizableResourceString(nameof(ThreadStaticOnNonStaticFieldDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        // [ThreadStatic]
+        // private static object t_field = new object();
+        internal static readonly DiagnosticDescriptor ThreadStaticInitializedInlineRule = DiagnosticDescriptorHelper.Create(ThreadStaticInitializedInlineRuleId,
+            CreateLocalizableResourceString(nameof(ThreadStaticInitializedInlineTitle)),
+            CreateLocalizableResourceString(nameof(ThreadStaticInitializedInlineMessage)),
+            DiagnosticCategory.Reliability,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(ThreadStaticInitializedInlineDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(ThreadStaticOnNonStaticFieldRule, ThreadStaticInitializedInlineRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(context =>
+            {
+                // Ensure ThreadStatic exists
+                if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemThreadStaticAttribute, out var threadStaticAttributeType))
+                {
+                    return;
+                }
+
+                // Warn on any [ThreadStatic] instance field.
+                context.RegisterSymbolAction(context =>
+                {
+                    ISymbol symbol = context.Symbol;
+
+                    // If it's a static symbol, nothing to see here.
+                    if (symbol.IsStatic)
+                    {
+                        return;
+                    }
+
+                    // If it's a field, it's the symbol we'll check.
+                    // If it's an auto-prop, find its backing field if there is one.
+                    IFieldSymbol? fieldSymbol = symbol as IFieldSymbol;
+                    if (symbol is IPropertySymbol)
```

Comment:
We have an extension method to determine if a symbol is a property with backing field: https://github.com/dotnet/roslyn-analyzers/blob/2b89a4115019ef1280706a6631b6b7c8f9840c9a/src/Utilities/Compiler/Extensions/ISymbolExtensions.cs#L110-L114

Consider adding an out `[NotNullWhen(true)] out IFieldSymbol? backingField` parameter to it and use it here.